library/x509write_csr.c

```diff
@@ -273,24 +280,33 @@ int mbedtls_x509write_csr_pem( mbedtls_x509write_csr *ctx, unsigned char *buf, s
                        int (*f_rng)(void *, unsigned char *, size_t),
                        void *p_rng )
 {
-    int ret;
-    unsigned char output_buf[4096];
+    int ret = 0;
+    unsigned char *output_buf;
+    enum { OUTPUT_BUF_LEN = 4096 };
     size_t olen = 0;
 
-    if( ( ret = mbedtls_x509write_csr_der( ctx, output_buf, sizeof(output_buf),
+    output_buf = mbedtls_calloc(1, OUTPUT_BUF_LEN);
```

Comment:
Since we're allocating memory dynamically, how about using a size based on `size` rather than a hard-coded constant? The size of the PEM data is 4/3 of the DER plus a few bytes plus the begin/end strings.